### PR TITLE
ARROW-5492: [R] Add "col_select" argument to read_* functions to read subset of columns

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(arrow, warn.conflicts = FALSE)
library(tibble)

tf <- tempfile()
write.csv(iris, tf, row.names = FALSE, quote = FALSE)

read_csv_arrow(tf, col_select = starts_with("Sepal"))
#> # A tibble: 150 x 2
#>    Sepal.Length Sepal.Width
#>           <dbl>       <dbl>
#>  1          5.1         3.5
#>  2          4.9         3  
#>  3          4.7         3.2
#>  4          4.6         3.1
#>  5          5           3.6
#>  6          5.4         3.9
#>  7          4.6         3.4
#>  8          5           3.4
#>  9          4.4         2.9
#> 10          4.9         3.1
#> # … with 140 more rows

tf <- tempfile()
write_feather(iris, tf)
read_feather(tf, col_select = contains("."))
#> # A tibble: 150 x 4
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width
#>           <dbl>       <dbl>        <dbl>       <dbl>
#>  1          5.1         3.5          1.4         0.2
#>  2          4.9         3            1.4         0.2
#>  3          4.7         3.2          1.3         0.2
#>  4          4.6         3.1          1.5         0.2
#>  5          5           3.6          1.4         0.2
#>  6          5.4         3.9          1.7         0.4
#>  7          4.6         3.4          1.4         0.3
#>  8          5           3.4          1.5         0.2
#>  9          4.4         2.9          1.4         0.2
#> 10          4.9         3.1          1.5         0.1
#> # … with 140 more rows
```

<sup>Created on 2019-06-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9000)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
only `read_feather()` and `read_csv_arrow()` for now.

**Comment by romainfrancois** [comment]
`read_feather()` and `read_parquet()` have direct support for `col_select=`

`read_csv_arrow()` and `read_json_arrow()` do it after the fact, i.e. they read the full `arrow::Table` and then do a `$select(<tidy selection>)` on it. 

This just exposes enough parquet api for this, I'll get to expose more parquet classes and methods in some other PR later.

File: r/src/parquet.cpp
Comment (romainfrancois) [inline r/src/parquet.cpp]: is it worth opening an issue about this, i.e. adding a `::arrow::Status GetSchema(std::shared_ptr<::arrow::Schema>* out);` method to `parquet::arrow::FileReader` ?

Comment (romainfrancois) [reply r/src/parquet.cpp]: https://github.com/apache/arrow/pull/4668

**Comment by romainfrancois** [comment]
The appveyor fail is unrelated

**Comment by nealrichardson** [comment]
The current Appveyor failure is related: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/25488050/job/x036oqg45dp67bo7#L3365

Until https://issues.apache.org/jira/browse/ARROW-5683 you'll need to `skip_on_os("windows")` any test that reads the parquet file included in `inst/`. See https://github.com/apache/arrow/pull/4627/files#diff-e5a0b5d08f0d85fde5678a5d99149739R23

File: r/configure
Comment (nealrichardson) [inline r/configure]: Please also remove `-DARROW_R_WITH_PARQUET` from configure.win

File: r/R/csv.R
Comment (nealrichardson) [inline r/R/csv.R]: Since this and read_json use $Read() then $select(), it would probably make sense to always mmap (https://issues.apache.org/jira/browse/ARROW-5502). Otherwise, IIUC, the readers will read the entire file into memory in Arrow before selecting the columns, which defeats the purpose of only selecting a few columns.

File: r/R/parquet.R
Comment (nealrichardson) [inline r/R/parquet.R]: Why does this reader have a `ReadTable` method while the others just have `Read`?

File: r/R/parquet.R
Comment (nealrichardson) [inline r/R/parquet.R]: This isn't very helpful documentation--can you elaborate, or at least point to the Arrow docs where they are enumerated?

File: r/src/parquet.cpp
Comment (nealrichardson) [reply r/src/parquet.cpp]: This PR isn't blocked by that, right? Looks like you could clean this up if that merges first but it's not contingent.

**Comment by nealrichardson** [comment]
https://github.com/apache/arrow/pull/4681 will fix the Appveyor failure without requiring the test to be skipped.

**Comment by nealrichardson** [comment]
Alright, #4681 merged so the Appveyor error should be fixed on rebase/merge upstream.

File: r/src/parquet.cpp
Comment (romainfrancois) [reply r/src/parquet.cpp]: It's not, I've used a workaround for now.

File: r/data-raw/codegen.R
Comment (romainfrancois) [reply r/data-raw/codegen.R]: everyone should have `pak` it's great. But ok I'll make the change.

File: r/R/parquet.R
Comment (romainfrancois) [reply r/R/parquet.R]: That's the name of the method in the C++ api:

https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader.h#L217

File: r/R/parquet.R
Comment (romainfrancois) [reply r/R/parquet.R]: At this point, these are really to check boxes, I suppose we can do a doc sweep at some point

File: r/R/csv.R
Comment (romainfrancois) [reply r/R/csv.R]: I've switched to `mmap_open()` but I don't think that makes much difference here anyway, the csv reader has no way as far as I understand to only read selected columns, so if we do want `col_select()` it's after the `Table` with all the columns is materialized anyway.

File: r/R/parquet.R
Comment (nealrichardson) [reply r/R/parquet.R]: Ok, I'll take on doc improvements after this merges.

File: r/R/csv.R
Comment (nealrichardson) [reply r/R/csv.R]: Alright, well, even so, I would expect the csv reader to grow better memory mapping support for this as the C++ Datasets API progresses, so we'll be ready for that.

**Review by nealrichardson — APPROVED** [review]
LGTM, let's merge this and I'll proceed with documentation cleanup after.

**Pull request closed without merging**
